File: speed/mathematical/mathematical.jl
using Random, LinearAlgebra, FFTW

# FFTW threading - use system default unless Windows causes issues
# Comment out the line below if you get segfaults on Windows
# FFTW.set_num_threads(1)

# If you need to disable FFTW threading due to crashes, uncomment this:
# FFTW.set_num_threads(1)

function matrix_operations(size::Int)::Float64
    Random.seed!(42)
    
    a = rand(Float64, size, size) .* 9.0 .+ 1.0
    b = rand(Float64, size, size) .* 9.0 .+ 1.0
    c = zeros(Float64, size, size)
    
    start_time = time_ns()
    
    # blocked matrix multiplication for cache efficiency
    block = 32
    for ii in 1:block:size
        for jj in 1:block:size
            for kk in 1:block:size
                i_end = min(ii + block - 1, size)
                j_end = min(jj + block - 1, size)
                k_end = min(kk + block - 1, size)
                
                @inbounds for i in ii:i_end
                    for j in jj:j_end
                        for k in kk:k_end
                            c[i, j] += a[i, k] * b[k, j]
                        end
                    end
                end
            end
        end
    end
    
    # matrix transpose
    temp = transpose(c)
    
    # matrix operations
    scalar = 1.5
    @inbounds for i in 1:size
        for j in 1:size
            c[i, j] = temp[i, j] + a[i, j] * scalar
        end
    end
    
    end_time = time_ns()
    
    # prevent optimization
    sum_diagonal = sum(@inbounds c[i, i] for i in 1:size)
    
    return (end_time - start_time) / 1_000_000
end

function is_prime_fast(n::Int64)::Bool
    if n < 2 return false end
    if n == 2 || n == 3 return true end
    if n % 2 == 0 || n % 3 == 0 return false end
    
    i = 5
    while i * i <= n
        if n % i == 0 || n % (i + 2) == 0
            return false
        end
        i += 6
    end
    return true
end

function factorize(n::Int)::Vector{Int}
    factors = Int[]
    i = 2
    while i * i <= n
        while n % i == 0
            push!(factors, i)
            n ÷= i
        end
        i += 1
    end
    if n > 1
        push!(factors, n)
    end
    return factors
end

function number_theory(limit::Int)::Float64
    start_time = time_ns()
    
    is_prime = trues(limit + 1)
    is_prime[1] = false
    if limit >= 2
        is_prime[2] = true
    end
    
    # segmented sieve - skip even numbers for efficiency
    i = 3
    while i * i <= limit
        if is_prime[i]
            j = i * i
            while j <= limit
                is_prime[j] = false
                j += i
            end
        end
        i += 2
    end
    
    # primality testing and factorization
    prime_count = 0
    composite_factors = 0
    
    start_range = max(1, limit - 1000)
    for i in start_range:limit
        if is_prime_fast(Int64(i))
            prime_count += 1
        else
            factors = factorize(i)
            composite_factors += length(factors)
        end
    end
    
    # twin prime counting  
    twin_primes = 0
    for i in 3:(limit-2)
        if is_prime[i] && is_prime[i + 2]
            twin_primes += 1
        end
    end
    
    end_time = time_ns()
    
    # prevent optimization
    result = prime_count + composite_factors + twin_primes
    
    return (end_time - start_time) / 1_000_000
end

function statistical_computing(samples::Int)::Float64
    start_time = time_ns()
    
    Random.seed!(42)
    
    # pre-allocate arrays for better performance
    inside_circle = 0
    values = Vector{Float64}(undef, samples)
    
    @inbounds for i in 1:samples
        x = rand()
        y = rand()
        if x^2 + y^2 <= 1.0
            inside_circle += 1
        end
        values[i] = randn()
    end
    
    pi_estimate = 4.0 * inside_circle / samples
    
    # statistical calculations - vectorized where possible
    mean_val = sum(values) / length(values)
    variance = sum((v - mean_val)^2 for v in values) / length(values)
    
    # numerical integration of sin(x) from 0 to π/2
    integration_samples = samples ÷ 4
    integral_sum = 0.0
    @inbounds for i in 1:integration_samples
        x = rand() * π / 2
        integral_sum += sin(x)
    end
    integral_result = (π / 2) * integral_sum / integration_samples
    
    end_time = time_ns()
    
    # prevent optimization
    result = pi_estimate + variance + integral_result
    
    return (end_time - start_time) / 1_000_000
end

function signal_processing(size::Int)::Float64
    Random.seed!(42)
    
    # create test signals with explicit types
    signal = Vector{ComplexF64}(undef, size)
    kernel = Vector{ComplexF64}(undef, size)
    
    @inbounds for i in 1:size
        signal[i] = complex(rand() * 2.0 - 1.0, rand() * 2.0 - 1.0)
        kernel[i] = complex(rand() * 2.0 - 1.0, 0.0)
    end
    
    start_time = time_ns()
    
    # forward fft - this should now use multiple threads if available
    signal_fft = fft(signal)
    kernel_fft = fft(kernel)
    
    # convolution in frequency domain
    result = signal_fft .* kernel_fft
    
    # inverse fft
    result = ifft(result)
    
    # round trip test for accuracy
    roundtrip = ifft(fft(signal))
    error_sum = sum(abs(roundtrip[i] - signal[i]) for i in 1:size)
    
    end_time = time_ns()
    
    # prevent optimization
    total_sum = sum(abs(val) for val in result) + error_sum
    
    return (end_time - start_time) / 1_000_000
end

function heapify!(arr::Vector{Int}, n::Int, i::Int)
    largest = i
    left = 2 * i
    right = 2 * i + 1
    
    if left <= n && arr[left] > arr[largest]
        largest = left
    end
    if right <= n && arr[right] > arr[largest]
        largest = right
    end
    
    if largest != i
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify!(arr, n, largest)
    end
end

function heap_sort!(arr::Vector{Int})
    n = length(arr)
    
    # build heap
    for i in n÷2:-1:1
        heapify!(arr, n, i)
    end
    
    # extract elements
    for i in n:-1:2
        arr[1], arr[i] = arr[i], arr[1]
        heapify!(arr, i-1, 1)
    end
end

function data_structures(size::Int)::Float64
    Random.seed!(42)
    
    data1 = rand(1:(size*10), size)
    data2 = collect(1:size)
    data3 = collect(size:-1:1)
    
    start_time = time_ns()
    
    # multiple sorting algorithms
    sort!(data1)
    heap_sort!(data2)
    sort!(data3)
    
    # merge operation - more efficient implementation
    merged = Vector{Int}(undef, size * 2)
    i, j, k = 1, 1, 1
    @inbounds while i <= size && j <= size
        if data1[i] <= data2[j]
            merged[k] = data1[i]
            i += 1
        else
            merged[k] = data2[j]
            j += 1
        end
        k += 1
    end
    @inbounds while i <= size
        merged[k] = data1[i]
        i += 1
        k += 1
    end
    @inbounds while j <= size
        merged[k] = data2[j]
        j += 1
        k += 1
    end
    
    # binary search operations
    found_count = 0
    @inbounds for _ in 1:2000
        target = rand(1:(size*10))
        if !isempty(searchsorted(data1, target))
            found_count += 1
        end
        if !isempty(searchsorted(data2, target))
            found_count += 1
        end
    end
    
    end_time = time_ns()
    
    # prevent optimization
    result = found_count + merged[size] + length(data3)
    
    return (end_time - start_time) / 1_000_000
end

function main()
    scale_factor = 1
    
    if length(ARGS) > 0
        try
            scale_factor = parse(Int, ARGS[1])
            if scale_factor < 1 || scale_factor > 5
                println(stderr, "Scale factor must be between 1 and 5")
                exit(1)
            end
        catch
            println(stderr, "Invalid scale factor: $(ARGS[1])")
            exit(1)
        end
    end
    
    total_time = 0.0
    
    # warm up Julia JIT compiler to avoid compilation overhead
    matrix_operations(10)
    number_theory(1000)
    statistical_computing(1000)
    signal_processing(64)
    data_structures(100)
    
    # actual benchmarks
    total_time += matrix_operations(40 * scale_factor)
    total_time += number_theory(80000 * scale_factor)
    total_time += statistical_computing(300000 * scale_factor)
    total_time += signal_processing(256 * scale_factor)
    total_time += data_structures(30000 * scale_factor)
    
    @printf "%.3f\n" total_time
end

# run main function
main()

File: speed/mathematical/mathematical.go
package main

import (
	"fmt"
	"math"
	"math/cmplx"
	"math/rand"
	"os"
	"sort"
	"strconv"
	"time"
)

func matrixOperations(size int) float64 {
	a := make([][]float64, size)
	b := make([][]float64, size)
	c := make([][]float64, size)
	temp := make([][]float64, size)
	
	for i := range a {
		a[i] = make([]float64, size)
		b[i] = make([]float64, size)
		c[i] = make([]float64, size)
		temp[i] = make([]float64, size)
	}
	
	rand.Seed(42)
	for i := 0; i < size; i++ {
		for j := 0; j < size; j++ {
			a[i][j] = rand.Float64()*9 + 1
			b[i][j] = rand.Float64()*9 + 1
		}
	}
	
	start := time.Now()
	
	// blocked matrix multiplication
	block := 32
	for ii := 0; ii < size; ii += block {
		for jj := 0; jj < size; jj += block {
			for kk := 0; kk < size; kk += block {
				iMax := min(ii+block, size)
				jMax := min(jj+block, size)
				kMax := min(kk+block, size)
				for i := ii; i < iMax; i++ {
					for j := jj; j < jMax; j++ {
						for k := kk; k < kMax; k++ {
							c[i][j] += a[i][k] * b[k][j]
						}
					}
				}
			}
		}
	}
	
	// matrix transpose
	for i := 0; i < size; i++ {
		for j := 0; j < size; j++ {
			temp[j][i] = c[i][j]
		}
	}
	
	// matrix operations
	scalar := 1.5
	for i := 0; i < size; i++ {
		for j := 0; j < size; j++ {
			c[i][j] = temp[i][j] + a[i][j]*scalar
		}
	}
	
	duration := time.Since(start)
	
	sum := 0.0
	for i := 0; i < size; i++ {
		sum += c[i][i]
	}
	_ = sum
	
	return float64(duration.Nanoseconds()) / 1000000.0
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func isPrimeFast(n int64) bool {
	if n < 2 {
		return false
	}
	if n == 2 || n == 3 {
		return true
	}
	if n%2 == 0 || n%3 == 0 {
		return false
	}
	
	for i := int64(5); i*i <= n; i += 6 {
		if n%i == 0 || n%(i+2) == 0 {
			return false
		}
	}
	return true
}

func factorize(n int) []int {
	factors := []int{}
	for i := 2; i*i <= n; i++ {
		for n%i == 0 {
			factors = append(factors, i)
			n /= i
		}
	}
	if n > 1 {
		factors = append(factors, n)
	}
	return factors
}

func numberTheory(limit int) float64 {
	start := time.Now()
	
	isPrime := make([]bool, limit+1)
	for i := range isPrime {
		isPrime[i] = true
	}
	isPrime[0] = false
	isPrime[1] = false
	
	// segmented sieve
	for i := 2; i*i <= limit; i++ {
		if isPrime[i] {
			for j := i * i; j <= limit; j += i {
				isPrime[j] = false
			}
		}
	}
	
	// primality testing and factorization
	primeCount := 0
	compositeFactors := 0
	for i := limit - 1000; i <= limit; i++ {
		if isPrimeFast(int64(i)) {
			primeCount++
		} else {
			factors := factorize(i)
			compositeFactors += len(factors)
		}
	}
	
	// twin prime counting
	twinPrimes := 0
	for i := 3; i <= limit-2; i++ {
		if isPrime[i] && isPrime[i+2] {
			twinPrimes++
		}
	}
	
	duration := time.Since(start)
	result := primeCount + compositeFactors + twinPrimes
	_ = result
	
	return float64(duration.Nanoseconds()) / 1000000.0
}

func statisticalComputing(samples int) float64 {
	start := time.Now()
	
	rand.Seed(42)
	insideCircle := 0
	values := make([]float64, 0, samples)
	
	// monte carlo and normal distribution sampling
	for i := 0; i < samples; i++ {
		x := rand.Float64()
		y := rand.Float64()
		if x*x+y*y <= 1.0 {
			insideCircle++
		}
		
		// box-muller for normal distribution
		if i%2 == 0 {
			u1 := rand.Float64()
			u2 := rand.Float64()
			z0 := math.Sqrt(-2*math.Log(u1)) * math.Cos(2*math.Pi*u2)
			values = append(values, z0)
		}
	}
	
	piEstimate := 4.0 * float64(insideCircle) / float64(samples)
	
	// statistical calculations
	mean := 0.0
	for _, val := range values {
		mean += val
	}
	mean /= float64(len(values))
	
	variance := 0.0
	for _, val := range values {
		diff := val - mean
		variance += diff * diff
	}
	variance /= float64(len(values))
	
	// numerical integration
	integrationSamples := samples / 4
	integralSum := 0.0
	for i := 0; i < integrationSamples; i++ {
		x := rand.Float64() * math.Pi / 2
		integralSum += math.Sin(x)
	}
	integralResult := (math.Pi / 2) * integralSum / float64(integrationSamples)
	
	duration := time.Since(start)
	result := piEstimate + variance + integralResult
	_ = result
	
	return float64(duration.Nanoseconds()) / 1000000.0
}

func fft(data []complex128) {
	n := len(data)
	if n <= 1 {
		return
	}
	
	even := make([]complex128, n/2)
	odd := make([]complex128, n/2)
	
	for i := 0; i < n/2; i++ {
		even[i] = data[i*2]
		odd[i] = data[i*2+1]
	}
	
	fft(even)
	fft(odd)
	
	for i := 0; i < n/2; i++ {
		t := cmplx.Exp(complex(0, -2*math.Pi*float64(i)/float64(n))) * odd[i]
		data[i] = even[i] + t
		data[i+n/2] = even[i] - t
	}
}

func ifft(data []complex128) {
	n := len(data)
	for i := range data {
		data[i] = cmplx.Conj(data[i])
	}
	fft(data)
	for i := range data {
		data[i] = cmplx.Conj(data[i]) / complex(float64(n), 0)
	}
}

func signalProcessing(size int) float64 {
	signal := make([]complex128, size)
	kernel := make([]complex128, size)
	result := make([]complex128, size)
	
	rand.Seed(42)
	for i := 0; i < size; i++ {
		real := rand.Float64()*2 - 1
		imag := rand.Float64()*2 - 1
		signal[i] = complex(real, imag)
		kernel[i] = complex(rand.Float64()*2-1, 0)
	}
	
	start := time.Now()
	
	// prepare fft data
	signalFFT := make([]complex128, size)
	kernelFFT := make([]complex128, size)
	copy(signalFFT, signal)
	copy(kernelFFT, kernel)
	
	// forward fft
	fft(signalFFT)
	fft(kernelFFT)
	
	// convolution in frequency domain
	for i := 0; i < size; i++ {
		result[i] = signalFFT[i] * kernelFFT[i]
	}
	
	// inverse fft
	ifft(result)
	
	// round trip test
	roundtrip := make([]complex128, size)
	copy(roundtrip, signal)
	fft(roundtrip)
	ifft(roundtrip)
	
	errorSum := 0.0
	for i := 0; i < size; i++ {
		errorSum += cmplx.Abs(roundtrip[i] - signal[i])
	}
	
	duration := time.Since(start)
	
	sum := 0.0
	for _, val := range result {
		sum += cmplx.Abs(val)
	}
	sum += errorSum
	_ = sum
	
	return float64(duration.Nanoseconds()) / 1000000.0
}

func heapify(arr []int, n, i int) {
	largest := i
	left := 2*i + 1
	right := 2*i + 2
	
	if left < n && arr[left] > arr[largest] {
		largest = left
	}
	if right < n && arr[right] > arr[largest] {
		largest = right
	}
	
	if largest != i {
		arr[i], arr[largest] = arr[largest], arr[i]
		heapify(arr, n, largest)
	}
}

func heapSort(arr []int) {
	n := len(arr)
	
	for i := n/2 - 1; i >= 0; i-- {
		heapify(arr, n, i)
	}
	
	for i := n - 1; i > 0; i-- {
		arr[0], arr[i] = arr[i], arr[0]
		heapify(arr, i, 0)
	}
}

func dataStructures(size int) float64 {
	data1 := make([]int, size)
	data2 := make([]int, size)
	data3 := make([]int, size)
	
	rand.Seed(42)
	for i := 0; i < size; i++ {
		data1[i] = rand.Intn(size*10) + 1
		data2[i] = i
		data3[i] = size - i
	}
	
	start := time.Now()
	
	// multiple sorting algorithms
	sort.Ints(data1)
	heapSort(data2)
	sort.Slice(data3, func(i, j int) bool { return data3[i] < data3[j] })
	
	// merge operation
	merged := make([]int, 0, size*2)
	i, j := 0, 0
	for i < len(data1) && j < len(data2) {
		if data1[i] <= data2[j] {
			merged = append(merged, data1[i])
			i++
		} else {
			merged = append(merged, data2[j])
			j++
		}
	}
	for i < len(data1) {
		merged = append(merged, data1[i])
		i++
	}
	for j < len(data2) {
		merged = append(merged, data2[j])
		j++
	}
	
	// binary search operations
	foundCount := 0
	for i := 0; i < 2000; i++ {
		target := rand.Intn(size*10) + 1
		idx1 := sort.SearchInts(data1, target)
		if idx1 < len(data1) && data1[idx1] == target {
			foundCount++
		}
		idx2 := sort.SearchInts(data2, target)
		if idx2 < len(data2) && data2[idx2] == target {
			foundCount++
		}
	}
	
	duration := time.Since(start)
	result := foundCount + len(merged) + len(data3)
	_ = result
	
	return float64(duration.Nanoseconds()) / 1000000.0
}

func main() {
	scaleFactor := 1
	
	if len(os.Args) > 1 {
		var err error
		scaleFactor, err = strconv.Atoi(os.Args[1])
		if err != nil {
			fmt.Println("Invalid scale factor:", os.Args[1])
			os.Exit(1)
		}
		if scaleFactor < 1 || scaleFactor > 5 {
			fmt.Println("Scale factor must be between 1 and 5")
			os.Exit(1)
		}
	}
	
	totalTime := 0.0
	
	totalTime += matrixOperations(40 * scaleFactor)
	totalTime += numberTheory(80000 * scaleFactor)
	totalTime += statisticalComputing(300000 * scaleFactor)
	totalTime += signalProcessing(256 * scaleFactor)
	totalTime += dataStructures(30000 * scaleFactor)
	
	fmt.Printf("%.3f\n", totalTime)
}